src/ekat/kokkos/ekat_kokkos_utils.hpp

```diff
@@ -64,10 +64,51 @@ void parallel_reduce (const TeamMember& team,
 
    result = local_tmp;
   } else {
-    Kokkos::parallel_reduce(Kokkos::TeamThreadRange(team, begin, end), lambda, result);
+    Kokkos::parallel_reduce(loop_boundaries, lambda, result);
   }
 }
 
+template <bool Serialize, int TotalSize, int VectorSize, typename TeamMember, typename InputProvider, typename ValueType>
+static KOKKOS_INLINE_FUNCTION
+void view_reduction (const TeamMember& team,
+                     const int& begin,
+                     const int& end,
+                     const InputProvider& input,
+                     ValueType& result)
+{
+  if (Serialize) {
+    const int  n_packs        = end - begin;
+    const bool has_garbage    = (TotalSize % VectorSize == 0 ? false : true);
+    const int  last_pack_size = (has_garbage ? TotalSize % VectorSize : VectorSize);
+    const int  loop_length    = (has_garbage ? n_packs-1 : n_packs);
+
+    impl::parallel_reduce<Serialize>(team,Kokkos::ThreadVectorRange(team, loop_length),
+                                     [&](const int k, ValueType& local_sum) {
+        local_sum = ekat::reduce_sum<Serialize>(input(k),local_sum);
+    }, result);
+      
+    // If last pack has garbage, we did not include it in the previous reduction,
+    // so manually add the last pack (only the non-garbage part)
+    if (has_garbage) {
+      Kokkos::single(Kokkos::PerTeam(team),[&] {
+        for (int j=0; j<last_pack_size; ++j) {
+          result += input(begin+n_packs-1)[j];
+        }
+      });
+    }
+  } else {
+    impl::parallel_reduce<Serialize>(team,Kokkos::ThreadVectorRange(team, TotalSize),
+                                     [&](const int k, ValueType& local_sum) {
```

Comment:
It could be helpful to implement the non-serialized case with packed logic. In hommexx, the if was on whether to do packed reduction or not. Maybe you wanted `if (!Serialize)` instead of `if (Serialize)'?

Reply:
Yes, I do have this backwards. The tests must have been passing/failing properly due to the `reduce_sum` function.